src/main/java/org/opentripplanner/routing/core/vehicle_sharing/VehiclePricingPackage.java

```diff
@@ -0,0 +1,243 @@
+        if (!isMaxRentingPriceUsed() || (isMaxRentingPriceUsed() && !isMaxRentingPriceExceeded(previousTotalPrice))) {
+            BigDecimal newTimeAssociatedPrice = BigDecimal.ZERO;
+
+            totalDrivingTimeInSeconds -= freeSeconds;
+
+            if (totalDrivingTimeInSeconds > 0) { //not all seconds of travel were free of charge
+                //Computing price associated with time ticks in package
+                int secondsInPackage = Math.min(totalDrivingTimeInSeconds, packageTimeLimitInSeconds);
+                BigDecimal timeTicksInPackage = BigDecimal.valueOf(secondsInPackage).divide(BigDecimal.valueOf(secondsPerTimeTickInPackage), 0, BigDecimal.ROUND_UP);
+                newTimeAssociatedPrice = timeTicksInPackage.multiply(drivingPricePerTimeTickInPackage);
+
+                //Computing price associated with time ticks above package
+                int secondsAbovePackage = totalDrivingTimeInSeconds-timeTicksInPackage.intValue()*secondsPerTimeTickInPackage;
+                if(secondsAbovePackage > 0){
+                    BigDecimal timeTicksAbovePackage = BigDecimal.valueOf(secondsAbovePackage).divide(BigDecimal.valueOf(secondsPerTimeTickInPackageExceeded), 0, BigDecimal.ROUND_UP);
+                    newTimeAssociatedPrice = newTimeAssociatedPrice.add(timeTicksAbovePackage.multiply(drivingPricePerTimeTickInPackageExceeded));
+                }
+            }
+            if (!isMaxRentingPriceUsed()){ //max renting time not used, no need to check anything else
+                return newTimeAssociatedPrice;
+            }
+            else{ //max renting price used, check whether it is going to be exceeded after adding price change
+                BigDecimal newTotalPrice = previousTotalPrice.subtract(currentTimePrice)
+                        .add(newTimeAssociatedPrice);
+                return isMaxRentingPriceExceeded(newTotalPrice)?maxRentingPrice.subtract(
+                        previousTotalPrice.subtract(currentTimePrice))
+                        :newTimeAssociatedPrice;
+            }
+        } else { //max renting price used and exceeded - do not increase price at this point
+            return currentTimePrice;
+        }
+    }
+
+    private boolean isMaxRentingPriceUsed(){
+        return maxRentingPrice.compareTo(BigDecimal.ZERO) == 1;
+    }
+
+    private boolean isMaxRentingPriceExceeded(BigDecimal totalVehiclePrice){
+        return totalVehiclePrice.compareTo(maxRentingPrice) >= 0;
+    }
+
+    public BigDecimal computeDistanceAssociatedPrice(BigDecimal currentStartPrice, BigDecimal currentTimePrice, BigDecimal currentDistancePrice, double totalDistanceInMeters){
+        BigDecimal previousTotalPrice = currentStartPrice.add(currentTimePrice).add(currentDistancePrice);
+        if (!isMaxRentingPriceUsed() || (isMaxRentingPriceUsed() && !isMaxRentingPriceExceeded(previousTotalPrice)) ||
```

Comment:
to się da uprościć do `if (!isMaxRentingPriceUsed() || !isMaxRentingPriceExceeded(previousTotalPrice) || kilometerPriceEnabledAboveMaxRentingPrice)`